sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/exemplar/FixedSizeExemplarReservoir.java

```diff
@@ -0,0 +1,182 @@
+ */
+public class FixedSizeExemplarReservoir implements ExemplarReservoir {
+  private final Clock clock;
+  private final ReservoirCell[] storage;
+  private final LongAdder numMeasurements = new LongAdder();
+
+  /**
+   * Instantiates an exemplar reservoir of fixed size.
+   *
+   * @param clock The clock to use when annotating measurements with time.
+   * @param size The number of exemplars to preserve.
+   */
+  public FixedSizeExemplarReservoir(Clock clock, int size) {
+    this.clock = clock;
+    this.storage = new ReservoirCell[size];
+    for (int i = 0; i < size; ++i) {
+      this.storage[i] = new ReservoirCell();
+    }
+  }
+
+  protected int maxSize() {
+    return storage.length;
+  }
+  /**
+   * Determines the sample bucket for a given measurement.
+   *
+   * @return The bucket to sample into or -1 for no sampling.
+   */
+  protected int bucketFor(double value, Attributes attributes, Context context) {
+    // TODO: hijackable random for testing.
+    long index = ThreadLocalRandom.current().nextLong(numMeasurements.sum() + 1);
+    if (index < storage.length) {
+      return (int) index;
+    }
+    return -1;
+  }
+
+  @Override
+  public void offerMeasurement(long value, Attributes attributes, Context context) {
+    offerMeasurement((double) value, attributes, context);
+  }
+
+  @Override
+  public void offerMeasurement(double value, Attributes attributes, Context context) {
+    int bucket = bucketFor(value, attributes, context);
+    if (bucket != -1) {
+      this.storage[bucket].offerMeasurement(value, attributes, context);
+    }
+    numMeasurements.increment();
+  }
+
+  @Override
+  public List<Exemplar> collectAndReset(Attributes pointAttributes) {
+    // Note: we are collecting exemplars from buckets piecemeal, but we
+    // could still be sampling exemplars during this process.
+    List<Exemplar> results = new ArrayList<>();
+    for (int i = 0; i < this.storage.length; ++i) {
+      Exemplar result = this.storage[i].getAndReset(pointAttributes);
+      if (result != null) {
+        results.add(result);
+      }
+    }
+    // Reset the count so exemplars are likely to be filled.
+    numMeasurements.reset();
+    return Collections.unmodifiableList(results);
+  }
+
+  /**
+   * A Reservoir cell pre-allocated memories for Exemplar data.
+   *
+   * <p>We only allocate new objects during collection. This class should NOT cause allocations
+   * during sampling or within the synchronous metric hot-path.
+   *
+   * <p>Allocations are acceptable in the {@link #getAndReset(Attributes)} method.
+   */
+  private class ReservoirCell {
+    private final Lock lock = new ReentrantLock();
+
+    private boolean hasValue = false;
+    private double value;
+    private Attributes attributes;
+    private String spanId;
+    private String traceId;
+    private long recordTime;
+
+    void offerMeasurement(double value, Attributes attributes, Context context) {
+      lock.lock();
```

Comment:
It's worth trying this with an old-school `synchronized(object)` approach. We found that approach to be significantly faster for some usages. Definitely worthy of a micro-benchmark.

Reply:
Noticing ~5ms improvement on that, going to update this with that change.   Still not happy with overall overhead of exemplars, but still chasing possible optimisations.